client/asset/btc/btc.go

```diff
@@ -2304,11 +2151,41 @@ func (btc *ExchangeWallet) run(ctx context.Context) {
 	}
 }
 
+// prepareRedemptionRequestsForBlockCheck prepares a copy of the
+// findRedemptionQueue, checking for missing block data along the way.
+func (btc *ExchangeWallet) prepareRedemptionRequestsForBlockCheck() []*findRedemptionReq {
+	// Search for contract redemption in new blocks if there
+	// are contracts pending redemption.
+	btc.findRedemptionMtx.RLock()
+	defer btc.findRedemptionMtx.RUnlock()
+	reqs := make([]*findRedemptionReq, 0, len(btc.findRedemptionQueue))
+	for _, req := range btc.findRedemptionQueue {
+		// If the request doesn't have a block hash yet, check if we can get one
+		// now.
+		if req.blockHash == nil {
+			btc.trySetRedemptionRequestBlock(req)
+		}
+		reqs = append(reqs, req)
+	}
+	return reqs
+}
+
 // checkForNewBlocks checks for new blocks. When a tip change is detected, the
 // tipChange callback function is invoked and a goroutine is started to check
 // if any contracts in the findRedemptionQueue are redeemed in the new blocks.
 func (btc *ExchangeWallet) checkForNewBlocks() {
-	newTipHash, err := btc.node.GetBestBlockHash()
+	reqs := btc.prepareRedemptionRequestsForBlockCheck()
+	// Redemption search would be compromised if the starting point cannot
+	// be determined, as searching just the new tip might result in blocks
+	// being omitted from the search operation. If that happens, cancel all
+	// find redemption requests in queue.
+	notifyFatalFindRedemptionError := func(s string, a ...interface{}) {
+		for _, req := range reqs {
+			req.fail("tipChange handler - "+s, a...)
+		}
+	}
```

Comment:
I think it would be good if this also canceled the searches. It looks like it can only notify the callers.

Reply:
It's not impossible that this request is also being run in another thread. We're now checking the condition with the helper function [here](https://github.com/buck54321/dcrdex/blob/f6ce73bc4908b41a2fd8487863212ae13dc308ef/client/asset/btc/btc.go#L1895). It's also of little consequence if another thread returns a result, since [subsequent results are simply ignored](https://github.com/buck54321/dcrdex/blob/f6ce73bc4908b41a2fd8487863212ae13dc308ef/client/asset/btc/btc.go#L419).